src/node/node_state.h

```diff
@@ -1481,15 +1483,37 @@ namespace ccf
           {
             for (auto& [k, v] : w)
             {
-              // If the version is not set (e.g. rekeying), use the version from
-              // the hook. Otherwise (e.g. recovery), use the version specified.
-              auto version_ =
-                v.value.wrapped_latest_ledger_secret.version == kv::NoVersion ?
-                version :
-                v.value.wrapped_latest_ledger_secret.version;
-
-              recovery_ledger_secrets.push_back(
-                {version_, v.value.encrypted_previous_ledger_secret});
+              kv::Version ledger_secret_version;
+              if (version == 1)
+              {
+                // Special case for the genesis transaction, which is applicable
+                // from the very first transaction
+                ledger_secret_version = 1;
+              }
+              else
+              {
+                // If the version is not set (rekeying), use the version
+                // from the hook plus one. Otherwise (recovery), use the
+                // version specified.
+                ledger_secret_version =
+                  v.value.wrapped_latest_ledger_secret.version ==
+                    kv::NoVersion ?
+                  (version + 1) :
+                  v.value.wrapped_latest_ledger_secret.version;
+              }
+
+              if (
```

Comment:
This check is necessary for the following reason. On a pure re-share (i.e. no new ledger secret created, for example when a new member is added), the `ccf.shares` table is now updated as follows:

```
Before:
ccf.shares: 0 -> [ (Latest_LS)k_z0; (Previous_LS)Latest_LS; (shares)k_z0 ]

After:
ccf.shares: 0 -> [ (Latest_LS)k_z1; None                  ; (shares)k_z1 ]
```

When replaying the ledger, the second value is not added to the `recovery_ledger_secrets` list as the ledger secrets were not updated in this case. When the recovery is finished and the members' shares have been combined, `Latest_LS` can be retrieved from the latest state and decrypted and the historical secrets decrypted in turn.

Reply:
I think this is worth a comment inline.